File: src/main/java/com/cuupa/classificator/services/kb/KnowledgeFileParser.java
package com.cuupa.classificator.services.kb;

import com.cuupa.classificator.services.kb.semantic.SenderToken;
import com.cuupa.classificator.services.kb.semantic.Topic;
import com.cuupa.classificator.services.kb.semantic.token.InvalidTokenException;
import com.cuupa.classificator.services.kb.semantic.token.MetaDataToken;
import com.cuupa.classificator.services.kb.semantic.token.TokenTextPointer;
import com.cuupa.classificator.services.kb.semantic.token.Tokens;
import org.apache.commons.lang3.tuple.Pair;
import org.jetbrains.annotations.NotNull;

public class KnowledgeFileParser {

    @NotNull
    static Topic parseTopicFile(@NotNull String kbFile) {
        validateToken(kbFile);
        return parseTopic(kbFile);
    }

    @NotNull
    static MetaDataToken parseMetaFile(@NotNull String kbFile) {
        validateToken(kbFile);
        return parseMetaData(kbFile);
    }

    @NotNull
    static SenderToken parseSenderFile(@NotNull String kbFile) {
        validateToken(kbFile);
        return parseSender(kbFile);
    }

    @NotNull
    private static SenderToken parseSender(@NotNull String kbFile) {
        String[] split = kbFile.split("=");
        String topicName = split[0].trim();

        SenderToken senderToken = new SenderToken();
        senderToken.setName(topicName);

        char[] charArray = kbFile.toCharArray();
        for (int index = 0; index < charArray.length; index++) {
            if (charArray[index] == '(') {
                senderToken.addToken(Tokens.get(new TokenTextPointer(charArray, index)));
            }
            if (charArray[index] == '}') {
                break;
            }
        }
        return senderToken;
    }

    @NotNull
    static Pair<String, String> parseRegexFile(@NotNull String filename, String content) {
        return Pair.of(filename.split("\\.")[0], content);
    }

    @NotNull
    private static MetaDataToken parseMetaData(@NotNull String kbFile) {
        MetaDataToken metadata = new MetaDataToken();
        char[] charArray = kbFile.toCharArray();
        for (int index = 0; index < charArray.length; index++) {
            if (charArray[index] == '$') {
                metadata.setName(findExtractName(charArray, index));
            } else if (charArray[index] == '(' && metadata.getName() != null && metadata.getName().length() > 0) {
                metadata.addToken(Tokens.get(new TokenTextPointer(charArray, index)));
            }
        }

        return metadata;
    }

    @NotNull
    public static Topic parseTopic(@NotNull String kbFile) {
        String[] split = kbFile.split("=");
        String topicName = split[0].trim();

        Topic topic = new Topic();
        topic.setName(topicName);

        char[] charArray = kbFile.toCharArray();
        for (int index = 0; index < charArray.length; index++) {
            if (charArray[index] == '(') {
                topic.addToken(Tokens.get(new TokenTextPointer(charArray, index)));
            }
            if (charArray[index] == '}') {
                break;
            }
        }

        // these are the metadata which is only applicable for the specific topic 
        if (kbFile.contains("$")) {
            MetaDataToken metadata = new MetaDataToken();
            for (int index = 0; index < charArray.length; index++) {
                if (charArray[index] == '$') {
                    metadata.setName(findExtractName(charArray, index));
                } else if (charArray[index] == '(' && metadata.getName() != null && metadata.getName().length() > 0) {
                    metadata.addToken(Tokens.get(new TokenTextPointer(charArray, index)));
                    topic.addMetaData(metadata);
                    metadata = new MetaDataToken();
                }
            }
        }

        return topic;
    }

    @NotNull
    private static String findExtractName(@NotNull final char[] charArray, int index) {
        StringBuilder extractName = new StringBuilder();
        for (int i = index; i < charArray.length; i++) {
            if (charArray[i] == '=') {
                for (int j = i - 1; j > 0; j--) {
                    if (charArray[j] == '$') {
                        return extractName.toString().trim();
                    }
                    extractName.insert(0, charArray[j]);
                }
            }
        }

        return extractName.toString().trim();
    }

    private static void validateToken(@NotNull String kbFile) {
        char[] charArray = kbFile.toCharArray();

        int curlyOpenBrackets = 0, curlyCloseBrackets = 0;
        int normalOpenBrackets = 0, normalCloseBrackets = 0;
        for (char c : charArray) {
            if (c == '{') {
                curlyOpenBrackets++;
            } else if (c == '}') {
                curlyCloseBrackets++;
            } else if (c == '(') {
                normalOpenBrackets++;
            } else if (c == ')') {
                normalCloseBrackets++;
            }
        }

        if (curlyCloseBrackets != curlyOpenBrackets || normalCloseBrackets != normalOpenBrackets) {
            throw new InvalidTokenException("invalid bracket count");
        }

        if (!kbFile.contains("=")) {
            throw new InvalidTokenException();
        }
    }
}

File: src/main/java/com/cuupa/classificator/services/kb/KnowledgeFileParser.kt
package com.cuupa.classificator.services.kb

import com.cuupa.classificator.constants.RegexConstants
import com.cuupa.classificator.services.kb.semantic.SenderToken
import com.cuupa.classificator.services.kb.semantic.Topic
import com.cuupa.classificator.services.kb.semantic.token.InvalidTokenException
import com.cuupa.classificator.services.kb.semantic.token.MetaDataToken
import com.cuupa.classificator.services.kb.semantic.token.TokenTextPointer
import com.cuupa.classificator.services.kb.semantic.token.Tokens
import org.apache.commons.lang3.tuple.Pair

object KnowledgeFileParser {

    fun parseTopicFile(kbFile: String): Topic {
        validateToken(kbFile)
        return parseTopic(kbFile)
    }

    fun parseMetaFile(kbFile: String): MetaDataToken {
        validateToken(kbFile)
        return parseMetaData(kbFile)
    }

    fun parseSenderFile(kbFile: String): SenderToken {
        validateToken(kbFile)
        return parseSender(kbFile)
    }

    private fun parseSender(kbFile: String): SenderToken {
        val senderToken = fillToken(kbFile)
        senderToken.name = kbFile.split(RegexConstants.equalPattern)[0].trim()
        return senderToken
    }

    private fun fillToken(kbFile: String): SenderToken {
        val token = SenderToken()
        val charArray = kbFile.toCharArray()
        for (index in charArray.indices) {
            if (charArray[index] == '(') {
                token.addToken(Tokens[TokenTextPointer(charArray, index)])
            }
            if (charArray[index] == '}') {
                break
            }
        }
        return token
    }

    fun parseRegexFile(filename: String, content: String): Pair<String, String> {
        return Pair.of(filename.split(RegexConstants.dotPattern)[0], content)
    }

    private fun parseMetaData(kbFile: String): MetaDataToken {
        val metadata = MetaDataToken()
        val charArray = kbFile.toCharArray()
        for (index in charArray.indices) {
            if (charArray[index] == '$') {
                metadata.name = findExtractName(charArray, index)
            } else if (charArray[index] == '(' && metadata.name.isNotEmpty()) {
                metadata.addToken(Tokens[TokenTextPointer(charArray, index)])
            }
        }
        return metadata
    }

    @JvmStatic
    fun parseTopic(kbFile: String): Topic {
        val split = kbFile.split(RegexConstants.equalPattern)
        val topicName = split[0].trim()
        val topic = Topic()
        topic.name = topicName
        val charArray = kbFile.toCharArray()
        for (index in charArray.indices) {
            if (charArray[index] == '(') {
                topic.addToken(Tokens[TokenTextPointer(charArray, index)])
            }
            if (charArray[index] == '}') {
                break
            }
        }
        return topic
    }

    private fun findExtractName(charArray: CharArray, index: Int): String {
        val extractName = StringBuilder()
        for (i in index until charArray.size) {
            if (charArray[i] == '=') {
                for (j in i - 1 downTo 1) {
                    if (charArray[j] == '$') {
                        return extractName.toString().trim { it <= ' ' }
                    }
                    extractName.insert(0, charArray[j])
                }
            }
        }
        return extractName.toString().trim()
    }

    private fun validateToken(kbFile: String) {
        val charArray = kbFile.toCharArray()
        var curlyOpenBrackets = 0
        var curlyCloseBrackets = 0
        var normalOpenBrackets = 0
        var normalCloseBrackets = 0
        for (c in charArray) {
            when (c) {
                '{' -> curlyOpenBrackets++
                '}' -> curlyCloseBrackets++
                '(' -> normalOpenBrackets++
                ')' -> normalCloseBrackets++

            }
        }
        if (curlyCloseBrackets != curlyOpenBrackets || normalCloseBrackets != normalOpenBrackets) {
            throw InvalidTokenException("invalid bracket count")
        }
        if (!kbFile.contains("=")) {
            throw InvalidTokenException("invalid file definition")
        }
    }
}